src/metabase/api/dashboard.clj

```diff
@@ -784,38 +785,50 @@
              :let  [{:keys [result_metadata]} card
+                    options  (:options ttag)
                     field-id (or
-                              ;; Get the field id from the field-clause if it contains it. This is the common case for
-                              ;; mbql queries.
-                              (mbql.u/match-one field-clause [:field (id :guard integer?) _] id)
-                              ;; Attempt to get the field clause from the model metadata corresponding to the field.
-                              ;; This is the common case for native queries in which mappings from original columns
-                              ;; have been performed using model metadata.
-                              (:id (qp.util/field->field-info field-clause result_metadata)))]
+                               ;; Get the field id from the field-clause if it contains it. This is the common case for
+                               ;; mbql queries.
+                               (mbql.u/match-one (:dimension ttag) [:field (id :guard integer?) _] id)
+                               ;; Attempt to get the field clause from the model metadata corresponding to the field.
+                               ;; This is the common case for native queries in which mappings from original columns
+                               ;; have been performed using model metadata.
+                               (:id (qp.util/field->field-info (:dimension ttag) result_metadata)))]
              :when field-id]
-         field-id)))
+         {:field-id field-id
+          :options  options})))
 
-(defn- param-key->field-ids
+(defn- param-key->param
   "Get Field ID(s) associated with a parameter in a Dashboard.
 
-    (param-key->field-ids (t2/select-one Dashboard :id 62) \"ee876336\")
+    (param-key->param (t2/select-one Dashboard :id 62) \"ee876336\")
     ;; -> #{276}"
   [dashboard param-key]
   {:pre [(string? param-key)]}
-  (let [{:keys [resolved-params]} (t2/hydrate dashboard :resolved-params)
-        param                     (get resolved-params param-key)]
-    (mappings->field-ids (:mappings param))))
+  (let [{:keys [resolved-params]} (t2/hydrate dashboard :resolved-params)]
+    (get resolved-params param-key)))
 
 (defn- chain-filter-constraints [dashboard constraint-param-key->value]
   (into {} (for [[param-key value] constraint-param-key->value
-                 field-id          (param-key->field-ids dashboard param-key)]
-             [field-id value])))
+                 :let              [param      (param-key->param dashboard param-key)
+                                    param-type (:type param)
+                                    op         (if (qualified-keyword? param-type)
+                                                 (keyword (name param-type))
+                                                 :=)]
+                 field             (mappings->fields (:mappings param))]
+             [(:field-id field)
```

Comment:
I don't think I like this data structure for constraints:
First, its length is not defined thus making it hard to destructure and understand what's inside. I'd rather see sth with this shape `[op value(s) option]`, this is guaranteed to have 3 parts and I know exactly what's what.

Second, it seems like the constraint here is a filter clause without the field because we're now sending both ops and value to `metabase.models.params.chain-filter/filter-clause`. reading [this comment](https://github.com/metabase/metabase/pull/34837/files#diff-1b120838a9ef31567ead6ca514337d29f0dc2da9428172057f4670624f53866aR123-R124) I think it's not the "official" API for the function, it's like someone tried to make a hack to fix some bugs in the past and sneak it in there.

the problem with that is it makes it hard to know what `value` is in the context of `filter-clause` function.

how about we restructure this even further and make constriant now have this schema.

```
[:map-of pos-int? [:map {:closed true}
                     [:value    :any]
                     [:op         [:enum := :=1! :between ....]]
                     [:options :map]]
```

with this we can update `filter-clause` to appropriately destructure then construct the filter clause.

Reply:
I'm not sure why having it as a list is better than a map of `field-id => constriantMap`.

Reply:
Well, for one it can have a few constraints on the same field without any weird dances. Plus map is a type where you'd be retrieving something from, generally, rather just iterating over - so it feels like a list/vector is a bit more honest about the intention.

Reply:
I see, I'm not totally against using list, but since the constraints are from `query-params` from the API, and it's already a map so I feel it a bit more natural to keep it that way.